macOS download_dir does not exist or is not writable, unless in verbose mode

Hi all,

I have constructor 3.1.0 from `conda install -y ctools/label/dev::constructor`
and when I try to create a new installer from the environment, I get the following error:
```
❯❯❯ constructor --cache-dir cache --output-dir dist .                                                                                                                           
platform: osx-64
Traceback (most recent call last):
  File "/Users/user/miniconda3/envs/test/bin/constructor", line 11, in <module>
    sys.exit(main())
  File "/Users/user/miniconda3/envs/test/lib/python3.8/site-packages/constructor/main.py", line 243, in main
    main_build(dir_path, output_dir=out_dir, platform=args.platform,
  File "/Users/user/miniconda3/envs/test/lib/python3.8/site-packages/constructor/main.py", line 111, in main_build
    fcp_main(info, verbose=verbose, dry_run=dry_run, conda_exe=conda_exe)
  File "/Users/user/miniconda3/envs/test/lib/python3.8/site-packages/constructor/fcp.py", line 383, in main
    _urls, dists, approx_tarballs_size, approx_pkgs_size, has_conda = _main(
  File "/Users/user/miniconda3/envs/test/lib/python3.8/site-packages/constructor/fcp.py", line 309, in _main
    pc_recs = _fetch(download_dir, precs)
  File "/Users/user/miniconda3/envs/test/lib/python3.8/site-packages/constructor/fcp.py", line 110, in _fetch
    assert pc.is_writable, download_dir + " does not exist or is not writable"
AssertionError: [---redacted---]/cache/osx-64 does not exist or is not writable
```

However, if I run it with the `-v` option, it does run successfully 😮 


I noticed that the setuid bit is set when using `-v`, but not otherwise.

```
❯❯❯ constructor -v --cache-dir cache --output-dir dist .
[...]

❯❯❯ ls -lha | grep cache
drwxrwsr-x   3 user  1244544504    96B Oct 31 21:24 cache

❯❯❯ rm -rf cache                                                                                                                                                             

❯❯❯ constructor --cache-dir cache --output-dir dist .
[...Error above...]

❯❯❯ ls -lha | grep cache
drwxr-x---   3 user  1244544504    96B Oct 31 21:24 cache

```

Any clues?

How to reproduce:
Install miniconda
```
./Miniconda3-latest-MacOSX-x86_64.sh

Welcome to Miniconda3 py38_4.8.3
...
```
 
open a new session:
```
❯❯❯  conda -V
conda 4.8.3

❯❯❯ conda create -y -n myenv python=3.8.5
❯❯❯ conda activate myenv
❯❯❯ conda install -y ctools/label/dev::constructor
❯❯❯ cat construct.yaml                                                                                                                                                      
name: test_installer
version: 1.0.0

channels:
  - http://repo.anaconda.com/pkgs/main/
  - https://repo.anaconda.com/pkgs/main/osx-64  [osx]
  - https://repo.anaconda.com/pkgs/main/noarch

environment: myenv
```

Thanks in advance.

--- a/constructor/conda_interface.py
+++ b/constructor/conda_interface.py
@@ -39,7 +39,6 @@ if conda_interface_type == 'conda':
     from conda.core.prefix_data import PrefixData as _PrefixData
     from conda.exports import default_prefix as _default_prefix
     from conda.models.channel import all_channel_urls as _all_channel_urls
-    from conda.gateways.disk import mkdir_p_sudo_safe as _mkdir_p_sudo_safe
     from conda.gateways.disk.read import read_paths_json as _read_paths_json
     from conda.models.dist import Dist as _Dist
     from conda.exports import MatchSpec as _MatchSpec
@@ -62,7 +61,6 @@ if conda_interface_type == 'conda':
     conda_context, env_vars = _conda_context, _env_vars
     conda_replace_context_default = _conda_replace_context_default
     download, PackageCacheRecord = _download, _PackageCacheRecord
-    mkdir_p_sudo_safe = _mkdir_p_sudo_safe
 
     # used by preconda.py
     Dist, MatchSpec, PrefixData, default_prefix = _Dist, _MatchSpec, _PrefixData, _default_prefix
@@ -143,6 +141,7 @@ if conda_interface_type == 'conda':
         # Choose an arbitrary old, expired date, so that conda will want to
         # immediately update it when not being run in offline mode
         url = used_repodata.pop('_url').rstrip("/")
+        used_repodata.pop("_mod", None)
         repodata = json.dumps(used_repodata, indent=2)
         repodata_header = json.dumps(
             {
@@ -154,3 +153,8 @@ if conda_interface_type == 'conda':
         repodata_filename = _cache_fn_url(url)
         with open(join(cache_dir, repodata_filename), 'w') as fh:
             fh.write(repodata)
+
+    def write_cache_dir():
+        cache_dir = join(PackageCacheData.first_writable().pkgs_dir, 'cache')
+        mkdir_p_sudo_safe(cache_dir)
+        return cache_dir

--- a/constructor/conda_interface.py
+++ b/constructor/conda_interface.py
@@ -39,6 +39,7 @@ if conda_interface_type == 'conda':
     from conda.core.prefix_data import PrefixData as _PrefixData
     from conda.exports import default_prefix as _default_prefix
     from conda.models.channel import all_channel_urls as _all_channel_urls
+    from conda.gateways.disk import mkdir_p_sudo_safe as _mkdir_p_sudo_safe
     from conda.gateways.disk.read import read_paths_json as _read_paths_json
     from conda.models.dist import Dist as _Dist
     from conda.exports import MatchSpec as _MatchSpec
@@ -61,6 +62,7 @@ if conda_interface_type == 'conda':
     conda_context, env_vars = _conda_context, _env_vars
     conda_replace_context_default = _conda_replace_context_default
     download, PackageCacheRecord = _download, _PackageCacheRecord
+    mkdir_p_sudo_safe = _mkdir_p_sudo_safe
 
     # used by preconda.py
     Dist, MatchSpec, PrefixData, default_prefix = _Dist, _MatchSpec, _PrefixData, _default_prefix

--- a/constructor/fcp.py
+++ b/constructor/fcp.py
@@ -21,7 +21,7 @@ from constructor.utils import hash_files, filename_dist
 from .conda_interface import (PackageCacheData, PackageCacheRecord, Solver, SubdirData,
                               VersionOrder, conda_context, conda_replace_context_default,
                               download, env_vars, read_paths_json, all_channel_urls,
-                              cc_platform, mkdir_p_sudo_safe)
+                              cc_platform)
 
 
 def getsize(filename):
@@ -106,10 +106,9 @@ platform: %(platform)s""" % dict(
 
 def _fetch(download_dir, precs):
     assert conda_context.pkgs_dirs[0] == download_dir
-    if not isdir(download_dir):
-        mkdir_p_sudo_safe(download_dir)
-    pc = PackageCacheData(download_dir)
-    assert pc.is_writable, download_dir + " does not exist or is not writable"
+    pc = PackageCacheData.first_writable()
+    assert pc.pkgs_dir == download_dir
+    assert pc.is_writable, f"{download_dir} does not exist or is not writable"
 
     for prec in precs:
         package_tarball_full_path = join(download_dir, prec.fn)

--- a/constructor/fcp.py
+++ b/constructor/fcp.py
@@ -21,7 +21,7 @@ from constructor.utils import hash_files, filename_dist
 from .conda_interface import (PackageCacheData, PackageCacheRecord, Solver, SubdirData,
                               VersionOrder, conda_context, conda_replace_context_default,
                               download, env_vars, read_paths_json, all_channel_urls,
-                              cc_platform)
+                              cc_platform, mkdir_p_sudo_safe)
 
 
 def getsize(filename):
@@ -107,7 +107,7 @@ platform: %(platform)s""" % dict(
 def _fetch(download_dir, precs):
     assert conda_context.pkgs_dirs[0] == download_dir
     if not isdir(download_dir):
-        os.makedirs(download_dir)
+        mkdir_p_sudo_safe(download_dir)
     pc = PackageCacheData(download_dir)
     assert pc.is_writable, download_dir + " does not exist or is not writable"
 

--- a/news/548-cache-creation
+++ b/news/548-cache-creation
@@ -1,0 +1,25 @@
+Enhancements:
+-------------
+
+* <news item>
+
+Bug fixes:
+----------
+
+* Freshly created download directories are now guaranteed to be writable (#411)
+
+Deprecations:
+-------------
+
+* <news item>
+
+Docs:
+-----
+
+* <news item>
+
+Other:
+------
+
+* <news item>
+

